feature/12.0/15510/replace_studio_project_fields

slsanc opened the pull request:
[ADD] Replace Studio Project Field with non-studio field

This commit works towards several related tasks, all of which require the replacement of the studio Project field with a non-studio field:

**[#12805 Auto populate Sales rep on Bill's based on PO [15505]](https://erp.bluestingray.com/web#id=15505&view_type=form&model=project.task&action=378&menu_id=263)**
**[#12816 Auto populate Sales rep on PO's based on Related Customer [15504]](https://erp.bluestingray.com/web#id=15504&view_type=form&model=project.task&action=378&menu_id=263)**
**[#12819 Expense App Auto Pop (less fields to enter) [15511]](https://erp.bluestingray.com/web#id=15511&view_type=form&model=project.task&action=378&menu_id=263)**
**[#12818 Auto Pop AA on SO & PO [15510]](https://erp.bluestingray.com/web#id=15510&view_type=form&model=project.task&action=378&menu_id=263)**

## Modules Added
This commit adds four modules:

- mesa_crm_leads_projects
- mesa_sale_autopopulate
- mesa_purchase_autopopulate
- mesa_expense_autopopulate

The mesa_crm_leads_projects module allows users to assign projects to CRM Leads. This is a functionality that Mesa added with Studio. Sale Autopopulate module calculates a Sale Order's Project to be the Project of its Opportunity. Because of this, that module depends on the Project field that mesa_crm_leads_projects adds to CRM Leads .

The Purchase app calculates a Purchase Order's Project to be the Project of the Purchase Order's Sale Order. Because of this, it depends on the Sale Autopopulate module to add a Project field to Sales Orders. The Expense Autopopulate module depends on the Sale Autopopulate module for similar reasons.

slsanc commented:
Some notes from previous PRs:

@traviswaelbro  asked:

> Can they still delete their Studio fields if we have them in our module? If not, then we need to either remove this field definition or replace the field

I checked and didn't get any issue with deleting the Project or Analytic studio fields, but I (predictably) get issues when I try to remove x_studio_related_customer from the Purchase Order model - the compute for Purchase Orders' Sales Rep relies upon that studio field.

> Can this be done without a force_user_id? I think the _inverse_user_id could just pass and this would work the same way. Just wondering if there's a way to do this without the extra force_ fields because they usually seem like overkill
I like the force_fields for a couple of reasons:

I plan on trying this tomorrow! Just a thought though - without a force_field as a second option, wouldn't the _compute_field(self) method just reset the field right after a user tries to change it?

manney commented on mesa_crm_lead_projects/models/crm/lead.py:
```diff
@@ -0,0 +1,37 @@
+
+        :return {NoneType}
+        """
+        for crm_lead in self:
+            if crm_lead.project_ids:
+                crm_lead.project_id = crm_lead.project_ids[0].id
```
Would something like this work?  The `filtered()` in the `for` cuts out the leads that don't have any projects.  The switch from `[0]` to `[:1]` just makes sure that there isn't an error if there isn't a record.

```suggestion
        for crm_lead in self.filtered('project_ids'):
            crm_lead.project_id = crm_lead.project_ids[:1].id
```

slsanc replied:
This works exactly the same - is there any performance benefit to this?

manney replied:
Potentially less iterations in the `for` loop.  If you do it the other way, then the `for` loop iterates over all of the leads whether they have projects or not.

manney commented on mesa_crm_lead_projects/models/project/project.py:
```diff
@@ -0,0 +1,29 @@
+class Project(models.Model):
+    _inherit = "project.project"
+
+    opportunity_id = fields.Many2one(
+        "crm.lead", ondelete="set null", string="Lead/Opportunity"
+    )
+
+    # The following is the current studio field for Lead/Opportunity.
+    x_studio_field_IxF50 = fields.Many2one(
+        "crm.lead",
+        compute="_compute_studio_opportunity",
+        string="Lead/Opportunity",
+        store=True,
+    )
+
+    @api.multi
+    @api.depends("opportunity_id")
+    def _compute_studio_opportunity(self):
+        """
+        When a user changes the non-studio "Lead/Opportunity" field, set the old
+        Studio field to the user's choice as well.
+
+        :return {NoneType}
+        """
+        for project_id in self:
+            project_id.x_studio_field_IxF50 = project_id.opportunity_id.id
```
I'm not even sure what you're doing here.  Why are we keeping `IxF50` around?

slsanc replied:
Because IxF50 is the current studio field for Opportunity, and other studio fields depend on it. If I were to leave this out, things that should change when the Opportunity is changed wouldn't.

One solution is to make all of those dependent fields non-studio, but that is a lot of work and is probably outside of the scope of these tasks. My solution is to set the studio field equal to the non-studio field.

manney commented on mesa_expense_autopopulate/models/hr/expense.py:
```diff
@@ -0,0 +1,160 @@
+        inverse="_inverse_warehouse",
+        store=True,
+    )
+    force_company_id = fields.Many2one("res.company", store=True)
+    force_project_id = fields.Many2one("project.project", store=True)
+    force_team_id = fields.Many2one("crm.team", store=True)
+    force_user_id = fields.Many2one("res.users", store=True)
+    force_warehouse_id = fields.Many2one("stock.warehouse", store=True)
```
Can we keep these in alphabetical format with the others, please?  I also want `string="<some name>"` on all of them too, please.

manney replied:
Sorry, can we also get the methods in alphabetical order too?  This will make everything easier to find.

slsanc replied:
@manney  do you mean that all the force_fields should come between company_id and project_id?

manney replied:
So:
```
company_id
force_company_id
force_project_id
force_team_id
force_user_id
force_warehouse_id
project_id
team_id
user_id
warehouse_id
...

manney replied:
At least that's how I doing in these days.  It makes a field/method easier to find.

manney commented on mesa_expense_autopopulate/models/hr/expense.py:
```diff
@@ -0,0 +1,160 @@
+        inverse="_inverse_company",
+        store=True,
+    )
+    project_id = fields.Many2one(
+        "project.project",
+        string="Project",
+        compute="_compute_project",
+        inverse="_inverse_project",
+    )
```
```suggestion
    project_id = fields.Many2one(
        "project.project",
        string="Project",
        compute="_compute_project",
        inverse="_inverse_project",
        store=True,
    )
```

manney commented on mesa_expense_autopopulate/models/hr/expense.py:
```diff
@@ -0,0 +1,160 @@
+        :return {NoneType}:
+        """
+        for expense_id in self:
+            expense_id.force_project_id = expense_id.x_studio_field_Evfct.id
```
What is going on here?
```suggestion
            expense_id.force_project_id = expense_id.project_id.id
```

manney commented on mesa_purchase_autopopulate/models/purchase/order.py:
```diff
@@ -0,0 +1,126 @@
+        string="Analytic Account",
+        store=True,
+    )
+    project_id = fields.Many2one(
+        "project.project", string="Project", related="sale_order_id.project_id"
+    )
```
You don't need a `string` if the related field is already named what you want it to be.  So if `sale_order_id.project_id`'s `string` is `Project` then you don't need to add a `string` here.

slsanc replied:
In Mesa's database, Sale Orders have no field 'project_id". they have a field "project_ids", which is a Many2many field, but they do not use this. They expect each Sale Order to list either one or zero Projects.

I will however remove the string for Analytic Account - I see this is redundant.

manney commented on mesa_purchase_autopopulate/models/purchase/order.py:
```diff
@@ -0,0 +1,126 @@
+    user_id = fields.Many2one(
+        compute="_compute_user_id", inverse="_inverse_user_id", store=True
+    )
+    force_analytic_account_id = fields.Many2one(
+        "account.analytic.account", string="Analytic Account (Custom)", store=True
+    )
+    force_user_id = fields.Many2one(
+        "res.users", string="Responsible Salesperson (Custom)", store=True
+    )
```
Again, please alphabetize these with the rest of the fields.

manney commented on mesa_purchase_autopopulate/models/purchase/order.py:
```diff
@@ -0,0 +1,126 @@
+                ]
+            else:
+                result["views"] = form_view
+                ### BEGIN CUSTOM CODE ###
+            if create_bill:
+                # If we want to create a Vendor Bill, Set the Purchase Order's
+                # Purchase Representative as the Vendor Bill's default
+                # Purchase Representative.
+                result["context"]["default_user_id"] = self.user_id.id
+                ### END CUSTOM CODE ###
+            else:
+                # Do not set an invoice_id if we want to create a new bill.
+                result["res_id"] = self.invoice_ids.id or False
```
I think this will help others see what you've done a bit better.  It took me a minute to compare it to Core to see what you did.  :slightly_smiling_face: 
```suggestion

            ### BEGIN CUSTOM CODE ###
            if create_bill:
                # If we want to create a Vendor Bill, Set the Purchase Order's
                # Purchase Representative as the Vendor Bill's default
                # Purchase Representative.
                result["context"]["default_user_id"] = self.user_id.id
            else:
                # Do not set an invoice_id if we want to create a new bill.
                # if not create_bill:
                ### END CUSTOM CODE ###
                result["res_id"] = self.invoice_ids.id or False

```

slsanc replied:
> Which PR were those notes from?

The Purchase Autopopulate Module.

manney commented on mesa_sale_autopopulate/models/sale/order.py:
```diff
@@ -0,0 +1,43 @@
+        store=True,
+    )
+    project_id = fields.Many2one(
+        "project.project", string="Project", related="opportunity_id.project_id"
```
Same comment about the `string` as above.

slsanc replied:
In Mesa's database, Sale Orders have no field 'project_id". they have a field "project_ids", which is a Many2many field, but they do not use this. They expect each Sale Order to list either one or zero Projects.

I will however remove the string for Analytic Account - I see this is redundant.

manney commented on mesa_sale_autopopulate/models/sale/order.py:
```diff
@@ -0,0 +1,43 @@
+    project_id = fields.Many2one(
+        "project.project", string="Project", related="opportunity_id.project_id"
+    )
+    force_analytic_account_id = fields.Many2one(
+        "account.analytic.account", string="Analytic Account (Custom)", store=True
+    )
```
Same comment about alphabetizing as above.

manney commented on vendor_pricelist/views/purchase_order.xml:
```diff
@@ -12,4 +12,4 @@
   </record> -->
-</odoo>
\ No newline at end of file
+</odoo>
```
Any reason why this entire file shouldn't be removed?

slsanc replied:
Apparently, that whole module was developed by Odoo. It will probably get deleted when they are no longer on Mesa.

manney reviewed (changes requested):
Got a few things in there to have another look at.

slsanc commented:
Update: I did in fact remove all of the force_fields. Functionality is unchanged. I also replaced Purchase Orders' Related Customer with a non-studio field. I also added the new non-studio field to the Purchase Order form view.

manney commented:
> Update: I did in fact remove all of the force_fields. Functionality is unchanged. I also replaced Purchase Orders' Related Customer with a non-studio field. I also added the new non-studio field to the Purchase Order form view.

Let's make sure we keep an eye on those.  I'm worried that a change in the value of a field that the compute depends on will wipe out the "custom" value or the value will be "lost" somehow.  I've seen it before on other projects.

manney commented on mesa_crm_lead_projects/models/crm/lead.py:
```diff
@@ -0,0 +1,36 @@
+class CrmLead(models.Model):
+    _inherit = "crm.lead"
+
+    project_ids = fields.One2many("project.project", "opportunity_id")
+    project_id = fields.Many2one(
+        "project.project", compute="_compute_project_id", string="Project"
+    )
+    project_count = fields.Integer(
+        compute="_compute_project_count", string="Project Card"
+    )
```
Sorry, missed this one.  Please alphabetize for clarity.

manney reviewed (changes requested):
A quick nitpick.

slsanc commented on mesa_purchase_autopopulate/models/purchase/order.py:
```diff
@@ -0,0 +1,122 @@
+        Compute the Purchase Order's Analytic Account. It defaults to the
+        Purchase Order's Project's Analytic Account.
+
+        :return {NoneType}:
+        """
+        for purchase_order in self:
+            purchase_order.analytic_account_id = (
+                purchase_order.project_id.analytic_account_id.id
+            )
+
+    @api.multi
+    @api.depends("customer_id")
+    def _compute_user_id(self):
+        """
+        Compute the Purchase Order's Sales Rep. It defaults to the Purchase
+        Order's Related Customer's Sales Rep.
+
+        :return {NoneType}:
+        """
+        for purchase_id in self:
+            purchase_id.user_id = purchase_id.customer_id.user_id.id
+
+    @api.multi
+    def _inverse_analytic_account(self):
+        """
+        This allows users to manually set the Analytic Account field in the
+        form.
+
+        :return {NoneType}:
+        """
+        pass
+
+    @api.multi
+    def _inverse_user_id(self):
+        """
+        This allows users to manually set the Sales Rep field in the form.
+
+        :return {NoneType}:
+        """
+        pass
```
I have tried replacing the computes with related fields - it caused several other things to break. I started getting security errors when trying to create certain types of records. I could spend some more time later trying to get it to work, but for now, this is working as expected.
